validator/client/src/main/java/tech/pegasys/artemis/validator/client/EpochDutiesScheduler.java

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.validator.client;
+
+import com.google.common.base.Throwables;
+import com.google.common.primitives.UnsignedLong;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import tech.pegasys.artemis.bls.BLSPublicKey;
+import tech.pegasys.artemis.datastructures.operations.Attestation;
+import tech.pegasys.artemis.datastructures.util.CommitteeUtil;
+import tech.pegasys.artemis.util.async.AsyncRunner;
+import tech.pegasys.artemis.util.async.SafeFuture;
+import tech.pegasys.artemis.validator.api.NodeSyncingException;
+import tech.pegasys.artemis.validator.api.ValidatorApiChannel;
+import tech.pegasys.artemis.validator.api.ValidatorDuties;
+import tech.pegasys.artemis.validator.client.duties.ScheduledDuties;
+
+class EpochDutiesScheduler {
+  private static final Logger LOG = LogManager.getLogger();
+  private final AsyncRunner asyncRunner;
+  private final ValidatorApiChannel validatorApiChannel;
+  private final ForkProvider forkProvider;
```

Comment:
Do we need to make sure to refresh the fork info when a reorg occurs?

Reply:
I don't think so.  The Fork data includes both "current" and "previous" forks and which epoch "current" takes effect.  Forks would be scheduled well in advance (and would require upgraded client software) so even if the reorg activated or deactivated a fork, the actual `Fork` data would still be the same - just a question of whether we are now before or after the appropriate epoch.